### ISSUE TITLE
Diff filter (patch) colors broken in 2.32

### ISSUE THREAD

**Issue by adriangoransson** [walles/riff#60]
With 2.32 the colorization of git commands using `-p`/`--patch` is broken.

2.31.1 on `master` (09c13be655614f8c5b503ff526b335693957049d).
<img width="835" alt="image" src="https://github.com/walles/riff/assets/4947727/8fa79c79-c7b6-4536-aad0-a612068e6f6d">
2.32.0
<img width="834" alt="image" src="https://github.com/walles/riff/assets/4947727/d1589664-a16b-4418-8586-cdff702d8e5a">

2.31.1 on 4d6c1fe29fae1a1385837d8185312824e0996613.
<img width="781" alt="image" src="https://github.com/walles/riff/assets/4947727/5e5c9d4f-5beb-4c0d-9663-6d534bf141d9">
2.32.0
<img width="787" alt="image" src="https://github.com/walles/riff/assets/4947727/be76eb29-51c4-455c-9a71-d8ef3702c412">

Tack för ett (i övrigt 😄) toppenbra verktyg!

**Comment by walles** [comment]
Here's a workaround while I'm thinking about this:
```
git config --global interactive.diffFilter "riff --color=on"
```

Caused by fixing #59, I'll have to think about how to make both of these work at the same time.

**Comment by walles** [comment]
I see some ways ahead:
* Drop the `--color=auto` mode and always color by default, but still let people turn it off if they want
* Make the workaround the official way for Git integration: `git config --global interactive.diffFilter "riff --color=on"`
* Detect that we're in interactive `git` mode somehow, and in that case turn on coloring in auto mode

Regarding detection, here's what Git sets in the environment of the diff filter (`riff`):
```sh
GIT_CONFIG_PARAMETERS='interactive.diffFilter'='/Users/johan/src/riff/hello.sh'
GIT_EXEC_PATH=/usr/local/opt/git/libexec/git-core
GIT_INDEX_FILE=.git/index
GIT_PREFIX=
```

So we can detect being launched from Git, but right now I can't see a way of detecting we're in interactive mode or that Git is also on the receiving end of our output.

FYI @alvaroaleman when adding `--color=auto` I unintentionally broke interactive Git operations.

**Comment by adriangoransson** [comment]
Thanks for the explanation, and workaround!

In an attempt to be helpful, I looked for some "prior work" on this topic by taking a quick look at Delta and diff-so-fancy.

- Delta has recommended the same workaround as you do since [2020 at least](https://github.com/dandavison/delta/commit/bc1ac13e92870ac89aba7417613edb819afe38d7).
- Diff-so-fancy also requires a flag (`--patch`) to make sure that the input/output matches. But even when I tried removing the flag (so that output is always matched), it does the right thing in regards to coloring the output. I haven't had the time to look any further, but I suspect that maybe they _always_ color the output?
